<h1>Create, modify, and delete columns</h1>

<small>Source: <a href="https://github.com/tidyverse/dplyr/blob/pkgdown-v1.1.4/R/mutate.R"><code>R/mutate.R</code></a></small>
<div><code>mutate.Rd</code></div>

<p><code>mutate()</code> creates new columns that are functions of existing variables. It can also modify (if the name is the same as an existing column) and delete columns (by setting their value to <code>NULL</code>).</p>

# <h2 id="ref-usage">Usage</h2>

<pre><code>mutate(.data, ...)

# S3 method for class 'data.frame'
mutate(
  .data,
  ...,
  .by = NULL,
  .keep = <a href="https://rdrr.io/r/base/c.html">c</a>("all", "used", "unused", "none"),
  .before = NULL,
  .after = NULL
)</code></pre>

# <h2 id="arguments">Arguments</h2>

<dl><dt id="arg--data">.data</dt>
<dd>
<p>A data frame, data frame extension (e.g. a tibble), or a
lazy data frame (e.g. from dbplyr or dtplyr). See <em>Methods</em>, below, for more details.</p>
</dd>

<dt id="arg--">...</dt>
<dd>
<p>&lt;<code><a href="https://rlang.r-lib.org/reference/args_data_masking.html">data-masking</a></code>&gt; Name-value pairs.
The name gives the name of the column in the output.</p>
<p>The value can be:</p><ul><li><p>A vector of length 1, which will be recycled to the correct length.</p></li>
<li><p>A vector the same length as the current group (or the whole data frame if ungrouped).</p></li>
<li><p><code>NULL</code>, to remove the column.</p></li>
<li><p>A data frame or tibble, to create multiple columns in the output.</p></li>
</ul>
</dd>

<dt id="arg--by">.by</dt>
<dd>
<p>&lt;<code><a href="https://dplyr.tidyverse.org/reference/dplyr_tidy_select.html">tidy-select</a></code>&gt; Optionally, a selection of columns to group by for just this operation, functioning as an alternative to <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code>. For details and examples, see <a href="https://dplyr.tidyverse.org/reference/dplyr_by.html">?dplyr_by</a>.</p>
</dd>

<dt id="arg--keep">.keep</dt>
<dd>
<p>Control which columns from <code>.data</code> are retained in the output. Grouping columns and columns created by <code>...</code> are always kept.</p>
<ul>
<li><p><code>"all"</code> retains all columns from <code>.data</code>. This is the default.</p></li>
<li><p><code>"used"</code> retains only the columns used in <code>...</code> to create new columns. This is useful for checking your work, as it displays inputs and outputs side-by-side.</p></li>
<li><p><code>"unused"</code> retains only the columns <em>not</em> used in <code>...</code> to create new columns. This is useful if you generate new columns, but no longer need the columns used to generate them.</p></li>
<li><p><code>"none"</code> doesn't retain any extra columns from <code>.data</code>. Only the grouping variables and columns created by <code>...</code> are kept.</p></li>
</ul>
</dd>

<dt id="arg--before-after">.before, .after</dt>
<dd>
<p>&lt;<code><a href="https://dplyr.tidyverse.org/reference/dplyr_tidy_select.html">tidy-select</a></code>&gt; Optionally, control where new columns should appear (the default is to add to the right hand side). See <code><a href="https://dplyr.tidyverse.org/reference/relocate.html">relocate()</a></code> for more details.</p>
</dd>
</dl>

# <h2 id="value">Value</h2>

<p>An object of the same type as <code>.data</code>. The output has the following properties:</p>

<ul>
<li><p>Columns from <code>.data</code> will be preserved according to the <code>.keep</code> argument.</p></li>
<li><p>Existing columns that are modified by <code>...</code> will always be returned in their original location.</p></li>
<li><p>New columns created through <code>...</code> will be placed according to the <code>.before</code> and <code>.after</code> arguments.</p></li>
<li><p>The number of rows is not affected.</p></li>
<li><p>Columns given the value <code>NULL</code> will be removed.</p></li>
<li><p>Groups will be recomputed if a grouping variable is mutated.</p></li>
<li><p>Data frame attributes are preserved.</p></li>
</ul>

# <h2 id="useful-mutate-functions">Useful mutate functions</h2>

<ul>
<li><p><code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code>, <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code>, <code><a href="https://rdrr.io/r/base/Log.html">log()</a></code>, etc., for their usual mathematical meanings</p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/lead-lag.html">lead()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/lead-lag.html">lag()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/row_number.html">dense_rank()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/row_number.html">min_rank()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/percent_rank.html">percent_rank()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/row_number.html">row_number()</a></code>,
<code><a href="https://dplyr.tidyverse.org/reference/percent_rank.html">cume_dist()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/ntile.html">ntile()</a></code></p></li>
<li><p><code><a href="https://rdrr.io/r/base/cumsum.html">cumsum()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/cumall.html">cummean()</a></code>, <code><a href="https://rdrr.io/r/base/cumsum.html">cummin()</a></code>, <code><a href="https://rdrr.io/r/base/cumsum.html">cummax()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/cumall.html">cumany()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/cumall.html">cumall()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/na_if.html">na_if()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/coalesce.html">coalesce()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/recode.html">recode()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when()</a></code></p></li>
</ul>

# <h2 id="grouped-tibbles">Grouped tibbles</h2>

<p>Because mutating expressions are computed within groups, they may
yield different results on grouped tibbles. This will be the case as soon as an aggregating, lagging, or ranking function is involved. Compare this ungrouped mutate:</p>

<pre><code>starwars %&gt;%
  select(name, mass, species) %&gt;%
  mutate(mass_norm = mass / mean(mass, na.rm = TRUE))</code></pre><p></p>

<p>With the grouped equivalent:</p>

<pre><code>starwars %&gt;%
  select(name, mass, species) %&gt;%
  group_by(species) %&gt;%
  mutate(mass_norm = mass / mean(mass, na.rm = TRUE))</code></pre><p></p>

<p>The former normalises <code>mass</code> by the global average whereas the latter normalises by the averages within species levels.</p>

# <h2 id="methods">Methods</h2>

<p>This function is a <strong>generic</strong>, which means that packages can provide implementations (methods) for other classes.
See the documentation of individual methods for extra arguments and differences in behaviour.</p>

<p>Methods available in currently loaded packages:
dbplyr (<code><a href="https://dbplyr.tidyverse.org/reference/mutate.tbl_lazy.html">tbl_lazy</a></code>), dplyr (<code>data.frame</code>)
.</p>

# <h2 id="see-also">See also</h2>

<div class="dont-index"><p>Other single table verbs:
<code><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a>()</code>,
<code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a>()</code>,
<code><a href="https://dplyr.tidyverse.org/reference/reframe.html">reframe</a>()</code>,
<code><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a>()</code>,
<code><a href="https://dplyr.tidyverse.org/reference/select.html">select</a>()</code>,
<code><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a>()</code>,
<code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a>()</code></p></div>

# <h2 id="ref-examples">Examples</h2>

In [2]:
library(dplyr)

# Newly created variables are available immediately
starwars |>
  select(name, mass) |>
  mutate(
    mass2 = mass * 2,
    mass2_squared = mass2 * mass2
  )


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




name,mass,mass2,mass2_squared
<chr>,<dbl>,<dbl>,<dbl>
Luke Skywalker,77.0,154.0,23716.00
C-3PO,75.0,150.0,22500.00
R2-D2,32.0,64.0,4096.00
Darth Vader,136.0,272.0,73984.00
Leia Organa,49.0,98.0,9604.00
Owen Lars,120.0,240.0,57600.00
Beru Whitesun Lars,75.0,150.0,22500.00
R5-D4,32.0,64.0,4096.00
Biggs Darklighter,84.0,168.0,28224.00


In [3]:
# As well as adding new variables, you can use mutate() to
# remove variables and modify existing variables.
starwars |>
  select(name, height, mass, homeworld) |>
  mutate(
    mass = NULL,
    height = height * 0.0328084 # convert to feet
  )

name,height,homeworld
<chr>,<dbl>,<chr>
Luke Skywalker,5.643045,Tatooine
C-3PO,5.479003,Tatooine
R2-D2,3.149606,Naboo
Darth Vader,6.627297,Tatooine
Leia Organa,4.921260,Alderaan
Owen Lars,5.839895,Tatooine
Beru Whitesun Lars,5.413386,Tatooine
R5-D4,3.182415,Tatooine
Biggs Darklighter,6.003937,Tatooine


In [4]:
# Use across() with mutate() to apply a transformation
# to multiple columns in a tibble.
starwars |>
  select(name, homeworld, species) |>
  mutate(across(!name, as.factor))

name,homeworld,species
<chr>,<fct>,<fct>
Luke Skywalker,Tatooine,Human
C-3PO,Tatooine,Droid
R2-D2,Naboo,Droid
Darth Vader,Tatooine,Human
Leia Organa,Alderaan,Human
Owen Lars,Tatooine,Human
Beru Whitesun Lars,Tatooine,Human
R5-D4,Tatooine,Droid
Biggs Darklighter,Tatooine,Human


In [5]:
# Window functions are useful for grouped mutates:
starwars |>
  select(name, mass, homeworld) |>
  group_by(homeworld) |>
  mutate(rank = min_rank(desc(mass)))

name,mass,homeworld,rank
<chr>,<dbl>,<chr>,<int>
Luke Skywalker,77.0,Tatooine,5
C-3PO,75.0,Tatooine,6
R2-D2,32.0,Naboo,6
Darth Vader,136.0,Tatooine,1
Leia Organa,49.0,Alderaan,2
Owen Lars,120.0,Tatooine,2
Beru Whitesun Lars,75.0,Tatooine,6
R5-D4,32.0,Tatooine,8
Biggs Darklighter,84.0,Tatooine,3


In [6]:
# By default, new columns are placed on the far right.
df <- tibble(x = 1, y = 2)
df |> mutate(z = x + y)

x,y,z
<dbl>,<dbl>,<dbl>
1,2,3


In [7]:
df |> mutate(z = x + y, .before = 1)

z,x,y
<dbl>,<dbl>,<dbl>
3,1,2


In [8]:
df |> mutate(z = x + y, .after = x)

x,z,y
<dbl>,<dbl>,<dbl>
1,3,2


In [9]:
# By default, mutate() keeps all columns from the input data.
df <- tibble(x = 1, y = 2, a = "a", b = "b")
df |> mutate(z = x + y, .keep = "all") # the default

x,y,a,b,z
<dbl>,<dbl>,<chr>,<chr>,<dbl>
1,2,a,b,3


In [10]:
df |> mutate(z = x + y, .keep = "used")

x,y,z
<dbl>,<dbl>,<dbl>
1,2,3


In [11]:
df |> mutate(z = x + y, .keep = "unused")

a,b,z
<chr>,<chr>,<dbl>
a,b,3


In [12]:
df |> mutate(z = x + y, .keep = "none")

z
<dbl>
3


In [13]:
# Grouping ----------------------------------------
# The mutate operation may yield different results on grouped
# tibbles because the expressions are computed within groups.
# The following normalises `mass` by the global average:
starwars |>
  select(name, mass, species) |>
  mutate(mass_norm = mass / mean(mass, na.rm = TRUE))

name,mass,species,mass_norm
<chr>,<dbl>,<chr>,<dbl>
Luke Skywalker,77.0,Human,0.7912704
C-3PO,75.0,Droid,0.7707179
R2-D2,32.0,Droid,0.3288397
Darth Vader,136.0,Human,1.3975685
Leia Organa,49.0,Human,0.5035357
Owen Lars,120.0,Human,1.2331487
Beru Whitesun Lars,75.0,Human,0.7707179
R5-D4,32.0,Droid,0.3288397
Biggs Darklighter,84.0,Human,0.8632041


In [14]:
# Whereas this normalises `mass` by the averages within species
# levels:
starwars |>
  select(name, mass, species) |>
  group_by(species) |>
  mutate(mass_norm = mass / mean(mass, na.rm = TRUE))

name,mass,species,mass_norm
<chr>,<dbl>,<chr>,<dbl>
Luke Skywalker,77.0,Human,0.9469930
C-3PO,75.0,Droid,1.0752688
R2-D2,32.0,Droid,0.4587814
Darth Vader,136.0,Human,1.6726110
Leia Organa,49.0,Human,0.6026319
Owen Lars,120.0,Human,1.4758332
Beru Whitesun Lars,75.0,Human,0.9223958
R5-D4,32.0,Droid,0.4587814
Biggs Darklighter,84.0,Human,1.0330833


<pre class="sourceCode r"><code>
# By default, new columns are placed on the far right.</span></span></span>
df</span> &lt;-</span> <a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span>(</span>x =</span> <span class="fl">1</span>, y =</span> <span class="fl">2</span>)</span></span></span>
df</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> mutate</span>(</span>z =</span> x</span> +</span> y</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       x     y     z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>     1     2     3</span>
df</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> mutate</span>(</span>z =</span> x</span> +</span> y</span>, .before =</span> <span class="fl">1</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       z     x     y</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>     3     1     2</span>
df</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> mutate</span>(</span>z =</span> x</span> +</span> y</span>, .after =</span> x</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       x     z     y</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>     1     3     2</span>
</span></span>
# By default, mutate() keeps all columns from the input data.</span></span></span>
df</span> &lt;-</span> <a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span>(</span>x =</span> <span class="fl">1</span>, y =</span> <span class="fl">2</span>, a =</span> "a"</span>, b =</span> "b"</span>)</span></span></span>
df</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> mutate</span>(</span>z =</span> x</span> +</span> y</span>, .keep =</span> "all"</span>)</span> # the default</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 5</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       x     y a     b         z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>     1     2 a     b         3</span>
df</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> mutate</span>(</span>z =</span> x</span> +</span> y</span>, .keep =</span> "used"</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       x     y     z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>     1     2     3</span>
df</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> mutate</span>(</span>z =</span> x</span> +</span> y</span>, .keep =</span> "unused"</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 3</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   a     b         z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> a     b         3</span>
df</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> mutate</span>(</span>z =</span> x</span> +</span> y</span>, .keep =</span> "none"</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 1</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span>     3</span>
</span></span>
# Grouping ----------------------------------------</span></span></span>
# The mutate operation may yield different results on grouped</span></span></span>
# tibbles because the expressions are computed within groups.</span></span></span>
# The following normalises `mass` by the global average:</span></span></span>
starwars</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span></span>
  <a href="select.html">select</a></span>(</span>name</span>, mass</span>, species</span>)</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span></span>
  mutate</span>(</span>mass_norm =</span> mass</span> /</span> <a href="https://rdrr.io/r/base/mean.html">mean</a></span>(</span>mass</span>, na.rm =</span> TRUE</span>)</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 87 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    name                mass species mass_norm</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Luke Skywalker        77 Human       0.791</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> C-3PO                 75 Droid       0.771</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> R2-D2                 32 Droid       0.329</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Darth Vader          136 Human       1.40 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Leia Organa           49 Human       0.504</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Owen Lars            120 Human       1.23 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Beru Whitesun Lars    75 Human       0.771</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> R5-D4                 32 Droid       0.329</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Biggs Darklighter     84 Human       0.863</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Obi-Wan Kenobi        77 Human       0.791</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 77 more rows</span></span>
</span></span>
# Whereas this normalises `mass` by the averages within species</span></span></span>
# levels:</span></span></span>
starwars</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span></span>
  <a href="select.html">select</a></span>(</span>name</span>, mass</span>, species</span>)</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span></span>
  <a href="group_by.html">group_by</a></span>(</span>species</span>)</span> <a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span></span>
  mutate</span>(</span>mass_norm =</span> mass</span> /</span> <a href="https://rdrr.io/r/base/mean.html">mean</a></span>(</span>mass</span>, na.rm =</span> TRUE</span>)</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 87 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Groups:   species [38]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    name                mass species mass_norm</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Luke Skywalker        77 Human       0.947</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> C-3PO                 75 Droid       1.08 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> R2-D2                 32 Droid       0.459</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Darth Vader          136 Human       1.67 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Leia Organa           49 Human       0.603</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Owen Lars            120 Human       1.48 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Beru Whitesun Lars    75 Human       0.922</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> R5-D4                 32 Droid       0.459</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Biggs Darklighter     84 Human       1.03 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Obi-Wan Kenobi        77 Human       0.947</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 77 more rows</span></span>
</span></span>
# Indirection ----------------------------------------</span></span></span>
# Refer to column names stored as strings with the `.data` pronoun:</span></span></span>
vars</span> &lt;-</span> <a href="https://rdrr.io/r/base/c.html">c</a></span>(</span>"mass"</span>, "height"</span>)</span></span></span>
mutate</span>(</span>starwars</span>, prod =</span> .data</span>[[</span>vars</span>[[</span><span class="fl">1</span>]</span>]</span>]</span>]</span> *</span> .data</span>[[</span>vars</span>[[</span><span class="fl">2</span>]</span>]</span>]</span>]</span>)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 87 × 15</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    name   height  mass hair_color skin_color eye_color birth_year sex  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Luke …    172    77 blond      fair       blue            19   male </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> C-3PO     167    75 <span style="color: #BB0000;">NA</span>         gold       yellow         112   none </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> R2-D2      96    32 <span style="color: #BB0000;">NA</span>         white, bl… red             33   none </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Darth…    202   136 none       white      yellow          41.9 male </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Leia …    150    49 brown      light      brown           19   fema…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Owen …    178   120 brown, gr… light      blue            52   male </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Beru …    165    75 brown      light      blue            47   fema…</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> R5-D4      97    32 <span style="color: #BB0000;">NA</span>         white, red red             <span style="color: #BB0000;">NA</span>   none </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Biggs…    183    84 black      light      brown           24   male </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Obi-W…    182    77 auburn, w… fair       blue-gray       57   male </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 77 more rows</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 7 more variables: gender &lt;chr&gt;, homeworld &lt;chr&gt;, species &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   films &lt;list&gt;, vehicles &lt;list&gt;, starships &lt;list&gt;, prod &lt;dbl&gt;</span></span>
# Learn more in ?rlang::args_data_masking</span></span></span>
</code></pre></div>